homeassistant/components/sensor/gitlab_ci.py

```diff
@@ -0,0 +1,182 @@
+        return self._state
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return {
+            ATTR_ATTRIBUTION: CONF_ATTRIBUTION,
+            ATTR_BUILD_STATUS: self._status,
+            ATTR_BUILD_STARTED: self._started_at,
+            ATTR_BUILD_FINISHED: self._finished_at,
+            ATTR_BUILD_DURATION: self._duration,
+            ATTR_BUILD_COMMIT_ID: self._commit_id,
+            ATTR_BUILD_COMMIT_DATE: self._commit_date,
+            ATTR_BUILD_ID: self._build_id,
+            ATTR_BUILD_BRANCH: self._branch
+        }
+
+    @property
+    def icon(self):
+        """Return the icon to use in the frontend."""
+        if self._status == 'success':
+            return ICON_HAPPY
+        if self._status == 'failed':
+            return ICON_SAD
+        return ICON_OTHER
+
+    def update(self):
+        """Collect updated data from GitLab API."""
+        self._gitlab_data.update()
+
+        self._status = self._gitlab_data.status
+        self._started_at = self._gitlab_data.started_at
+        self._finished_at = self._gitlab_data.finished_at
+        self._duration = self._gitlab_data.duration
+        self._commit_id = self._gitlab_data.commit_id
+        self._commit_date = self._gitlab_data.commit_date
+        self._build_id = self._gitlab_data.build_id
+        self._branch = self._gitlab_data.branch
+        self._state = self._gitlab_data.status
+
+
+class GitLabData():
+    """GitLab Data object."""
+
+    def __init__(self, gitlab_id, priv_token, interval):
+        """Fetch json data from GitLab API for most recent CI job."""
+        self._gitlab_id = gitlab_id
+        self._private_access_token = {'PRIVATE-TOKEN': priv_token}
+        self._url = ("https://gitlab.com/api/v4/projects/" +
+                     self._gitlab_id + "/jobs?per_page=1&page=1")
+        self._interval = interval
+        self.update = Throttle(interval)(self._update)
+
+        self._response = None
+        self._response_json = None
+        self.status = STATE_UNKNOWN
+        self.started_at = STATE_UNKNOWN
+        self.finished_at = STATE_UNKNOWN
+        self.duration = STATE_UNKNOWN
+        self.commit_id = STATE_UNKNOWN
+        self.commit_date = STATE_UNKNOWN
+        self.build_id = STATE_UNKNOWN
+        self.branch = STATE_UNKNOWN
+        self.state = STATE_UNKNOWN
+
+    def _update(self):
+        _logger = logging.getLogger(__name__)
+        _logger.debug(self._interval)
+        import requests
+        import json
+        try:
+            response = requests.get(self._url,
+                                    headers=self._private_access_token)
+            response.raise_for_status()
+            self._response = response.text[1:-1]
+            self._response_json = json.loads(self._response)
+            self._response_json = json.loads(self._response)
+
+            self.status = self._response_json['status']
```

Comment:
Please extract this into a third party lib.

Reply:
I've located an active 3rd party lib, https://github.com/python-gitlab/python-gitlab
I'll do a re-write over the next few days to a week or so to integrate this instead of using a json parse, as requested.

Reply:
Refactored to use `python-gitlab` 3rd party lib. Added as a `requirement`